kibana-reports/server/routes/utils/reportHelper.ts

```diff
@@ -95,58 +102,104 @@ export const createVisualReport = async (
+  const timePending = Date.now();
+  const saveParams: RequestParams.Index = {
+    index: CONFIG_INDEX_NAME.report,
+    id: savedReportId,
+    body: {
+      ...report,
+      state: REPORT_STATE.pending,
+      time_created: timePending,
+    },
+  };
 
-  if (reportSource === REPORT_TYPE.savedSearch) {
-    // TODO: Add createDataReport(report)
-    console.log('placeholder for createDataReport');
-  } else if (
-    reportSource === REPORT_TYPE.dashboard ||
-    reportSource === REPORT_TYPE.visualization
-  ) {
-    createReportResult = await createVisualReport(report);
+  const esResp = await callCluster(client, 'index', saveParams);
+  const reportId = esResp._id;
+
+  const reportDefinition = report.report_definition;
+  const reportParams = reportDefinition.report_params;
+  const reportSource = reportParams.report_source;
+  try {
+    if (reportSource === REPORT_TYPE.savedSearch) {
+      // TODO: Add createDataReport(report)
+      console.log('placeholder for createDataReport');
+    } else if (
+      reportSource === REPORT_TYPE.dashboard ||
+      reportSource === REPORT_TYPE.visualization
+    ) {
+      const queryUrl = report.query_url;
+      createReportResult = await createVisualReport(reportParams, queryUrl);
+    }
+  } catch (error) {
+    // update report instance with "error" state
+    const timeError = Date.now();
+    const updateParams: RequestParams.Update = {
+      id: reportId,
+      index: CONFIG_INDEX_NAME.report,
+      body: {
+        doc: {
+          state: REPORT_STATE.error,
+          time_created: timeError,
+        },
+      },
+    };
+
+    await callCluster(client, 'update', updateParams);
+
+    throw error;
   }
 
-  // save report instance with created state
-  // TODO: save report instance with error state
-  report = {
-    ...report,
-    time_created: createReportResult.timeCreated,
-    state: REPORT_STATE.created,
+  // update report document with state "created" and time_created
+  const updateParams: RequestParams.Update = {
+    id: reportId,
+    index: CONFIG_INDEX_NAME.report,
+    body: {
+      doc: {
+        time_created: createReportResult.timeCreated,
+        state: REPORT_STATE.created,
+      },
+    },
   };
 
-  const params: RequestParams.Index = {
-    index: 'report',
-    body: report,
-  };
-
-  //@ts-ignore
-  await client.callAsInternalUser('index', params);
+  await callCluster(client, 'update', updateParams);
 
   return createReportResult;
 };
 
-function getFileName(itemName: string, timeCreated: string): string {
-  return `${itemName}_${timeCreated}_${uuidv1()}`;
+function getFileName(itemName: string, timeCreated: Date): string {
+  return `${itemName}_${timeCreated.toISOString()}_${uuidv1()}`;
 }
+
+const callCluster = async (
+  client: IClusterClient | IScopedClusterClient,
+  endpoint: string,
+  params: any
+) => {
+  let esResp;
+  if ('callAsCurrentUser' in client) {
+    esResp = await client.callAsCurrentUser(endpoint, params);
```

Comment:
CallAsCurrentUser uses the credential from http request, while call as internal user uses the credential from kibana.yml file. Since executor and generate report request handler share the same logic, poller/ executor is like a background process, so there is no user request, we have to use callAsInternalUser https://github.com/elastic/kibana/blob/master/docs/development/core/server/kibana-plugin-core-server.iscopedclusterclient.md